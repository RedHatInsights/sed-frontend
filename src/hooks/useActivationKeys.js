import { useQuery } from '@tanstack/react-query';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';

const fetchActivationKeysData = (token) => async () => {
  const response = await fetch('/api/rhsm/v2/activation_keys', {
    headers: { Authorization: `Bearer ${await token}` },
  });

  const activationKeysData = await response.json();

  return activationKeysData.body;
};

const getActivationKeys = (token) => async () => {
  const keysData = await fetchActivationKeysData(token)();
  return keysData;
};

const useActivationKeys = () => {
  const chrome = useChrome();

  return useQuery(['activation_keys'], getActivationKeys(chrome?.auth?.getToken()));
};

export { useActivationKeys as default };
