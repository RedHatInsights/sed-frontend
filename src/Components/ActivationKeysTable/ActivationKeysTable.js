import React from 'react';
import { Link, useLocation } from 'react-router-dom';
import {
  Table /* data-codemods */,
  Thead,
  Tr,
  Th,
  Tbody,
  Td,
} from '@patternfly/react-table';
import useActivationKeys from '../../hooks/useActivationKeys';
import Loading from '../LoadingState/Loading';
import Unavailable from '@redhat-cloud-services/frontend-components/Unavailable';
import DeleteActivationKeyButton from '../ActivationKeys/DeleteActivationKeyButton';
import PropTypes from 'prop-types';

const ActivationKeysTable = (props) => {
  const { onDelete } = props;
  const columnNames = {
    name: 'Key Name',
    role: 'Role',
    serviceLevel: 'SLA',
    usage: 'Usage',
  };
  const { isLoading, error, data } = useActivationKeys();
  const location = useLocation();

  const Results = () => {
    return (
      <Table aria-label="ActivationKeys">
        <Thead>
          <Tr ouiaSafe={true}>
            <Th width={40}>{columnNames.name}</Th>
            <Th>{columnNames.role}</Th>
            <Th>{columnNames.serviceLevel}</Th>
            <Th>{columnNames.usage}</Th>
            <Td></Td>
          </Tr>
        </Thead>
        <Tbody>
          {data.map((datum) => {
            return (
              <Tr key={datum.name} ouiaSafe={true}>
                <Td modifier="breakWord" dataLabel={columnNames.name}>
                  <Link to={`${location.pathname}/${datum.name}`}>
                    {' '}
                    {datum.name}
                  </Link>
                </Td>
                <Td dataLabel={columnNames.role}>{datum.role}</Td>
                <Td dataLabel={columnNames.serviceLevel}>
                  {datum.serviceLevel}
                </Td>
                <Td dataLabel={columnNames.usage}>{datum.usage}</Td>
                <Td>
                  <DeleteActivationKeyButton
                    onClick={() => onDelete(datum.name)}
                  />
                </Td>
              </Tr>
            );
          })}
        </Tbody>
      </Table>
    );
  };

  if (isLoading && !error) {
    return <Loading />;
  } else if (!isLoading && !error) {
    return <Results />;
  } else {
    return <Unavailable />;
  }
};

ActivationKeysTable.propTypes = {
  onDelete: PropTypes.func.isRequired,
};

export default ActivationKeysTable;
