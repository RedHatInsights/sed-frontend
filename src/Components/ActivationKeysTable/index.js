import ActivationKeysTable from './ActivationKeysTable';

export { ActivationKeysTable as default };
