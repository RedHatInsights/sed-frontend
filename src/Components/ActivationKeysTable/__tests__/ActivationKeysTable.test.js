import React from 'react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import ActivationKeysTable from '../ActivationKeysTable';
import useActivationKeys from '../../../hooks/useActivationKeys';
import { get, def } from 'bdd-lazy-var';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
jest.mock('../../../hooks/useActivationKeys');
jest.mock('uuid', () => {
  return { v4: jest.fn(() => '00000000-0000-0000-0000-000000000000') };
});
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({ pathname: '/connector/test-key' }),
}));

const queryClient = new QueryClient();

const Table = () => (
  <BrowserRouter>
    <QueryClientProvider client={queryClient}>
      <ActivationKeysTable onDelete={() => {}} />
    </QueryClientProvider>
  </BrowserRouter>
);

jest.mock(
  '@redhat-cloud-services/frontend-components/Unavailable',
  // eslint-disable-next-line react/display-name
  () => () => <div>Unavailable</div>
);

describe('ActivationKeysTable', () => {
  def('rbacPermissions', () => {
    return {
      rbacPermissions: {
        canReadActivationKeys: true,
        canWriteActivationKeys: true,
      },
    };
  });
  def('loading', () => false);
  def('error', () => false);
  def('data', () => [
    {
      name: 'A',
      role: 'B',
      serviceLevel: 'C',
      usage: 'D',
    },
  ]);

  beforeEach(() => {
    jest
      .spyOn(queryClient, 'getQueryData')
      .mockReturnValue(get('rbacPermissions'));
    useActivationKeys.mockReturnValue({
      isLoading: get('loading'),
      error: get('error'),
      data: get('data'),
    });
  });

  it('renders correctly', () => {
    const { container } = render(<Table />);

    expect(container).toMatchSnapshot();
  });

  describe('when loading', () => {
    def('loading', () => true);

    it('renders the loading state', () => {
      const { container } = render(<Table />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('when user does not have write permissions', () => {
    def('rbacPermissions', () => {
      return {
        rbacPermissions: {
          canReadActivationKeys: true,
          canWriteActivationKeys: false,
        },
      };
    });
  });

  describe('when there is an error', () => {
    def('error', () => true);

    it('renders the error state', () => {
      const { container } = render(<Table />);

      expect(container).toMatchSnapshot();
    });
  });
});
