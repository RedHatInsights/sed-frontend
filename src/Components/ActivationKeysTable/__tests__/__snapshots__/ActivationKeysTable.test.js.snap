// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivationKeysTable renders correctly 1`] = `
<div>
  <table
    aria-label="ActivationKeys"
    class="pf-v5-c-table pf-m-grid-md"
    data-ouia-component-id="OUIA-Generated-Table-1"
    data-ouia-component-type="PF5/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class="pf-v5-c-table__thead"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="OUIA-Generated-TableRow-1"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <th
          class="pf-v5-c-table__th pf-m-width-40"
          scope="col"
          tabindex="-1"
        >
          Key Name
        </th>
        <th
          class="pf-v5-c-table__th"
          scope="col"
          tabindex="-1"
        >
          Role
        </th>
        <th
          class="pf-v5-c-table__th"
          scope="col"
          tabindex="-1"
        >
          SLA
        </th>
        <th
          class="pf-v5-c-table__th"
          scope="col"
          tabindex="-1"
        >
          Usage
        </th>
        <td
          class="pf-v5-c-table__td"
          tabindex="-1"
        />
      </tr>
    </thead>
    <tbody
      class="pf-v5-c-table__tbody"
      role="rowgroup"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="OUIA-Generated-TableRow-2"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td pf-m-break-word"
          data-label="Key Name"
          tabindex="-1"
        >
          <a
            href="/connector/test-key/A"
          >
             
            A
          </a>
        </td>
        <td
          class="pf-v5-c-table__td"
          data-label="Role"
          tabindex="-1"
        >
          B
        </td>
        <td
          class="pf-v5-c-table__td"
          data-label="SLA"
          tabindex="-1"
        >
          C
        </td>
        <td
          class="pf-v5-c-table__td"
          data-label="Usage"
          tabindex="-1"
        >
          D
        </td>
        <td
          class="pf-v5-c-table__td"
          tabindex="-1"
        >
          <div
            style="display: contents;"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ActivationKeysTable when loading renders the loading state 1`] = `
<div>
  <div
    class="pf-v5-l-bullseye"
  >
    <svg
      aria-label="Contents"
      aria-valuetext="Loading..."
      class="pf-v5-c-spinner pf-m-xl"
      role="progressbar"
      viewBox="0 0 100 100"
    >
      <circle
        class="pf-v5-c-spinner__path"
        cx="50"
        cy="50"
        fill="none"
        r="45"
      />
    </svg>
  </div>
</div>
`;

exports[`ActivationKeysTable when there is an error renders the error state 1`] = `
<div>
  <div>
    Unavailable
  </div>
</div>
`;
